src/goto-harness/goto_harness_parse_options.cpp

```diff
@@ -7,46 +7,164 @@ Author: Diffblue Ltd.
 \******************************************************************/
 
 #include <cstddef>
-#include <iostream>
+#include <set>
 #include <string>
+#include <utility>
 
+#include <goto-programs/goto_convert.h>
+#include <goto-programs/goto_model.h>
+#include <goto-programs/read_goto_binary.h>
+#include <goto-programs/write_goto_binary.h>
+#include <util/exception_utils.h>
 #include <util/exit_codes.h>
+#include <util/invariant.h>
 #include <util/version.h>
 
+#include "function_harness_generator.h"
+#include "goto_harness_generator_factory.h"
 #include "goto_harness_parse_options.h"
 
+// The basic idea is that this module is handling the following
+// sequence of events:
+// 1. Initialise a goto-model by parsing an input (goto) binary
+// 2. Initialise the harness generator (with some config) that will handle
+//    the mutation of the goto-model.
+// 3. Write the end result of that process to the output binary
+
 int goto_harness_parse_optionst::doit()
 {
   if(cmdline.isset("version"))
   {
-    std::cout << CBMC_VERSION << '\n';
+    status() << CBMC_VERSION << '\n';
     return CPROVER_EXIT_SUCCESS;
   }
 
-  help();
-  return CPROVER_EXIT_USAGE_ERROR;
+  auto factory = make_factory();
+  auto config = handle_common_options();
+
+  // Read goto binary into goto-model
+  auto read_goto_binary_result =
+    read_goto_binary(config.in_file, get_message_handler());
+  if(!read_goto_binary_result.has_value())
+  {
+    throw deserialization_exceptiont{"failed to read goto program from file `" +
+                                     config.in_file + "'"};
+  }
+  auto goto_model = std::move(read_goto_binary_result.value());
+
+  // Initialise harness generator
+  auto harness_generator = factory.factory(config.harness_type);
```

Comment:
I assume that when you create the harness_generator here, it will (potentially) do a bunch of checks for commandline/configuration values - which may potentially result in the generator failing to be constructed and emitting a user error message (e.g. if it doesn't understand how to configure itself from the users options). If so, is it better to move this before the goto_binary is read? e.g. get all the user command line processing out of the way before attempting to actually load the specified goto_binary?

Reply:
@chrisr-diffblue The option checking is to be done in `invoke_harness` (misnomer, name change pending). Basically, the idea here was to explicitly iterate over all arguments that _aren't_ handled by goto-harness itself, and force each generator to either handle it and throw an exception if it can't, so we're not doing that in the constructor (note that we're not passing the commandline to the factory here, and we expect that the builder for each generator shouldn't "cheat" and grab its arguments either)